### PR TITLE
[Feature][Transaction] Support two phase commit (2PC) for stream load

### LINKED ISSUE #7141: [Feature] [Stream Load] Two-Phase Commit for stream load
### Background
In the sample Doris application, data flow is as follows:
* read streaming data from Kafka
* Execute ETL in Flink
* Sink data batch to Doris by `stream load`

Flink generates checkpoints on a regular, configurable interval and then writes the checkpoint to a persistent storage system, such as HDFS. A checkpoint in Flink is a consistent snapshot of:
* The current state of an application
* The consumption progress of data stream(`offset`)

2021-11-17 20-43-16 的屏幕截图

In the event of a machine or Flink software failure and upon restart, the Flink application resumes processing from the most recent successfully-completed checkpoint, which causes partial data to be loaded to Doris twice and duplicate data.

To provide exactly-once semantics, Doris must provide a means to commit or rollback load that coordinate with Flink's checkpoints. So, it's better to support `Two-Phase Commit(2PC)` for stream load.

For the data sink to provide exactly-once guarantees, it must:
* write all data to Doris through several stream load tasks between two checkpoints (All data is non-visible).
* commit all stream load tasks between two checkpoints(All data is visible).

In the event of a machine or Flink software failure and upon restart, commit all stream load tasks between the most recent two checkpoints(It is ok to execute commit repeatedly for a stream load task).

Reference: https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html

### Design

The design of the two phase for stream load is as follows:

* First Phase:

2021-11-02 15-55-26 的屏幕截图

* Second Phase:
2021-11-02 15-57-47 的屏幕截图

Once the `pre-commit` is complete, we must ensure that the `commit` can be successful. Of course, Data releated to expired transaction which has been pre-committed could be removed.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by weizuo93**
## Proposed changes

The two phase batch commit means：
During Stream load, after data is written, the message will be returned to the client, the data is invisible at this point and the transaction status is PRECOMMITTED. The data will be visible only after COMMIT is triggered by client.
    
1. User can invoke the following interface to trigger commit operations for transaction：

```
    curl -X PUT --location-trusted -u user:passwd -H "txn_id:txnId" -H "txn_operation:commit" http://fe_host:http_port/api/{db}/_stream_load_2pc
```

or

```
    curl -X PUT --location-trusted -u user:passwd -H "txn_id:txnId" -H "txn_operation:commit" http://be_host:webserver_port/api/{db}/_stream_load_2pc
```
    
2.User can invoke the following interface to trigger abort operations for transaction：

```
    curl -X PUT --location-trusted -u user:passwd -H "txn_id:txnId" -H "txn_operation:abort" http://fe_host:http_port/api/{db}/_stream_load_2pc
```

or

```
    curl -X PUT --location-trusted -u user:passwd -H "txn_id:txnId" -H "txn_operation:abort" http://be_host:webserver_port/api/{db}/_stream_load_2pc
```

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #7141 ) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

**Comment by morningman** [comment]
Thanks for your contribution. Please rebase the master to pass the test first.

File: fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/LoadAction.java [hunk 1]
```diff
@@ -73,10 +78,29 @@ public Object streamLoad(HttpServletRequest request,
...
+        this.isStreamLoad = true;
+        executeCheckPassword(request, response);
+        return executeStreamLoadAbort(request, response, db, table);
+    }
+
     // Same as Multi load, to be compatible with http v1's response body,
     // we return error by using RestBaseResult.
     private Object executeWithoutPassword(HttpServletRequest request,
                                           HttpServletResponse response, String db, String table) {
+        LOG.info("received stream load request (precommit  httpv2).");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/LoadAction.java:103]: This log can be removed.

File: fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/LoadAction.java [hunk 2]
```diff
@@ -154,4 +178,125 @@ private Object executeWithoutPassword(HttpServletRequest request,
...
+            }
+
+            String fullDbName = ClusterNamespace.getFullName(clusterName, dbName);
+
+            List<Long> transactionIds = Lists.newArrayList();
+            String txnIds = request.getHeader(TXN_KEY);
+            if (Strings.isNullOrEmpty(txnIds)) {
+                return new RestBaseResult("No transaction id selected.");
+            } else {
+                for (String txnId : txnIds.split(",")) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/LoadAction.java:208]: trim the space?

File: fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/LoadAction.java [hunk 3]
```diff
@@ -154,4 +178,125 @@ private Object executeWithoutPassword(HttpServletRequest request,
...
+            // check auth
+            checkTblAuth(ConnectContext.get().getCurrentUserIdentity(), fullDbName, tableName, PrivPredicate.LOAD);
+            LOG.info("redirect stream load commit request to master FE, txns: {}", txnIds);
+
+            RedirectView redirectView = redirectToMaster(request, response);
+            if (redirectView != null) {
+                return redirectView;
+            }
+            {
+                LOG.info("Master FE received http request to commit txns: {}", txnIds);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/LoadAction.java:222]: Use `debug` level, or there will be too many logs

File: fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java [hunk 4]
```diff
@@ -342,6 +339,344 @@ public void updateDatabaseUsedQuotaData(long usedQuotaDataBytes) {
...
+                    }
+                }
+            }
+        }
+
+        unprotectedPreCommitTransaction(transactionState, errorReplicaIds, tableToPartition, totalInvolvedBackends, db);
+        LOG.info("transaction:[{}] successfully pre-committed", transactionState);
+    }
+
+    public TransactionState checkPreCommitStatus(long transactionId) throws UserException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java:516]:
```suggestion
    private TransactionState checkPreCommitStatus(long transactionId) throws UserException {
```

File: fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java [hunk 5]
```diff
@@ -342,6 +339,344 @@ public void updateDatabaseUsedQuotaData(long usedQuotaDataBytes) {
...
+                }
+
+                if (transactionState != null) {
+                    transactionState.setCheckTimeout(true);
+                }
+            }
+        }
+    }
+
+    public void commitTransaction(TransactionState transactionState) throws UserException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java:665]:
```suggestion
    private void commitTransaction(TransactionState transactionState) throws UserException {
```

File: fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java [hunk 6]
```diff
@@ -342,6 +339,344 @@ public void updateDatabaseUsedQuotaData(long usedQuotaDataBytes) {
         this.usedQuotaDataBytes = usedQuotaDataBytes;
     }
 
+    public void preCommitTransaction(List<Table> tableList, long transactionId, List<TabletCommitInfo> tabletCommitInfos,
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java:342]: Most part of this method is same as `commitTransaction`. Can we merge these 2 methods?
I think the only difference it to set the txn state to PRECOMMITTED or COMMITTED.

File: fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java [hunk 7]
```diff
@@ -198,6 +224,16 @@ public void commitTransaction(long dbId, List<Table> tableList, long transaction
         DatabaseTransactionMgr dbTransactionMgr = getDatabaseTransactionMgr(dbId);
         dbTransactionMgr.commitTransaction(tableList, transactionId, tabletCommitInfos, txnCommitAttachment);
     }
+
+    public void commitTransaction(long dbId, List<Long> transactionIds)
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:228]:
```suggestion
    private void commitTransaction(long dbId, List<Long> transactionIds)
```

Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:228]: Better to change the method name, for example, adding a prefix of suffix `2PC` to make it different from the normal `commitTransaction`. Or it is confused for code reading.

All other methods which are related to `2PC` in both GlobalTransactionMgr and DatabaseTransactionMgr need to be changed.

File: fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java [hunk 8]
```diff
@@ -342,6 +339,344 @@ public void updateDatabaseUsedQuotaData(long usedQuotaDataBytes) {
...
+        if (transactionState.getTransactionStatus() == TransactionStatus.PREPARE) {
+            LOG.debug("transaction is prepare, not pre-committed: {}", transactionId);
+            throw new TransactionCommitFailedException("transaction [" + transactionId
+                    + "] is prepare, not pre-committed");
+        }
+
+        long currentTimeMillis = System.currentTimeMillis();
+        // Maybe new  invisible version has not been reported to master FE after the transaction is pre-committed
+        // or FE restart, we should wait a little while to commit transaction.
+        if (currentTimeMillis - transactionState.getPreCommitTime() <
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java:560]: Does it mean that we need to wait at least 60sec to commit the precommitted the txn?
That is strange to me.

File: fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java [hunk 9]
```diff
@@ -342,6 +339,344 @@ public void updateDatabaseUsedQuotaData(long usedQuotaDataBytes) {
...
+        transactionState.setErrorReplicas(errorReplicaIds);
+
+        return transactionState;
+    }
+
+    public void commitTransaction(List<Long> transactionIds) throws UserException {
+        try {
+            List<TransactionState> transactionStates = Lists.newArrayList();
+            for (long transactionId : transactionIds) {
+                transactionStates.add(checkPreCommitStatus(transactionId));
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java:642]: If the txn is in PRECOMMITTED, it should be able to be COMMITTED, why we still need to call `checkPreCommitStatus()` here?
The only thing need to be done in `checkPreCommitStatus()` is to check whether the txn state is PRECOMMITTED, but I saw that there other checking list in `checkPreCommitStatus()`, why?

File: be/src/http/action/stream_load_2pc.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    std::string req_txn_id = req->header(HTTP_TXN_ID_KEY);
+    try {
+        ctx->txn_id = std::stoull(req_txn_id);
+    } catch (const std::exception& e) {
+        status = Status::InternalError("convert txn_id [" + req_txn_id + "] failed");
+        status_result = to_json(status);
+        HttpChannel::send_reply(req, HttpStatus::OK, status_result);
+        return;
+    }
+    ctx->txn_operation = req->header(HTTP_TXN_OPERATION_KEY);
```
Comment (morningman) [inline be/src/http/action/stream_load_2pc.cpp:63]: check invalid parameter?

File: be/src/common/config.h [hunk 11]
```diff
@@ -353,6 +353,7 @@ CONF_mInt32(stream_load_record_batch_size, "50");
 CONF_Int32(stream_load_record_expire_time_secs, "28800");
 // time interval to clean expired stream load records
 CONF_mInt64(clean_stream_load_record_interval_secs, "1800");
+CONF_mBool(disable_stream_load_2pc, "false");
```
Comment (morningman) [inline be/src/common/config.h:356]: I think we can set default value to true. And set it true after we complete the flink connector and everything is ok,

File: gensrc/thrift/FrontendService.thrift [hunk 12]
```diff
@@ -656,6 +656,23 @@ struct TLoadTxnCommitResult {
     1: required Status.TStatus status
 }
 
+struct TLoadTxn2PCRequest {
+    1: optional string cluster
+    2: required string user
```
Comment (morningman) [inline gensrc/thrift/FrontendService.thrift:661]: use `optional` for all fields

File: be/src/http/action/stream_load_2pc.cpp (hunk 10 above)
Comment (weizuo93) [reply be/src/http/action/stream_load_2pc.cpp:63]: > check invalid parameter?

OK, thank you.

File: be/src/common/config.h (hunk 11 above)
Comment (weizuo93) [reply be/src/common/config.h]: > I think we can set default value to true. And set it true after we complete the flink connector and everything is ok,

OK.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**